src/lib.rs

```diff
@@ -74,544 +82,81 @@ mod nl;
-// const SIOCGSTAMP: c_int = 0x8906;
-
-// get timestamp in a struct timespec (ns accuracy)
-const SIOCGSTAMPNS: c_int = 0x8907;
-
-/// if set, indicate 29 bit extended format
-pub const EFF_FLAG: u32 = 0x80000000;
-
-/// remote transmission request flag
-pub const RTR_FLAG: u32 = 0x40000000;
-
-/// error flag
-pub const ERR_FLAG: u32 = 0x20000000;
-
-/// valid bits in standard frame id
-pub const SFF_MASK: u32 = 0x000007ff;
-
-/// valid bits in extended frame id
-pub const EFF_MASK: u32 = 0x1fffffff;
-
-/// valid bits in error frame
-pub const ERR_MASK: u32 = 0x1fffffff;
-
-
-/// an error mask that will cause SocketCAN to report all errors
-pub const ERR_MASK_ALL: u32 = ERR_MASK;
-
-/// an error mask that will cause SocketCAN to silently drop all errors
-pub const ERR_MASK_NONE: u32 = 0;
-
-fn c_timeval_new(t: time::Duration) -> timeval {
-    timeval {
-        tv_sec: t.as_secs() as time_t,
-        tv_usec: t.subsec_micros() as suseconds_t,
-    }
-}
-
-#[derive(Debug)]
-#[repr(C)]
-struct CanAddr {
-    _af_can: c_short,
-    if_index: c_int, // address familiy,
-    rx_id: u32,
-    tx_id: u32,
-}
-
-/// A socket for a CAN device.
-///
-/// Will be closed upon deallocation. To close manually, use std::drop::Drop.
-/// Internally this is just a wrapped file-descriptor.
-#[derive(Debug)]
-pub struct CanSocket {
-    fd: c_int,
-}
-
-impl CanSocket {
-    /// Open a named CAN device.
-    ///
-    /// Usually the more common case, opens a socket can device by name, such
-    /// as "vcan0" or "socan0".
-    pub fn open(ifname: &str) -> Result<CanSocket, CanSocketOpenError> {
-        let if_index = if_nametoindex(ifname)?;
-        CanSocket::open_if(if_index)
-    }
-
-    /// Open CAN device by interface number.
-    ///
-    /// Opens a CAN device by kernel interface number.
-    pub fn open_if(if_index: c_uint) -> Result<CanSocket, CanSocketOpenError> {
-        let addr = CanAddr {
-            _af_can: AF_CAN as c_short,
-            if_index: if_index as c_int,
-            rx_id: 0, // ?
-            tx_id: 0, // ?
-        };
-
-        // open socket
-        let sock_fd;
-        unsafe {
-            sock_fd = socket(PF_CAN, SOCK_RAW, CAN_RAW);
-        }
-
-        if sock_fd == -1 {
-            return Err(CanSocketOpenError::from(io::Error::last_os_error()));
-        }
-
-        // bind it
-        let bind_rv;
-        unsafe {
-            let sockaddr_ptr = &addr as *const CanAddr;
-            bind_rv = bind(sock_fd,
-                           sockaddr_ptr as *const sockaddr,
-                           size_of::<CanAddr>() as u32);
-        }
-
-        // FIXME: on fail, close socket (do not leak socketfds)
-        if bind_rv == -1 {
-            let e = io::Error::last_os_error();
-            unsafe {
-                close(sock_fd);
-            }
-            return Err(CanSocketOpenError::from(e));
-        }
-
-        Ok(CanSocket { fd: sock_fd })
-    }
+impl embedded_hal::can::nb::Can for CanSocket {
```

Comment:
Why are these sitting out here, rather than in the `socket` module? Is this the stuff that would be moved upstream into the embedded-hal/can crate?